File: internal/todoist/todoist.go
package todoist

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"

	"github.com/ivanlee1999/sesh/internal/db"
)

const apiBase = "https://api.todoist.com/rest/v2"

type Task struct {
	ID        string `json:"id"`
	Content   string `json:"content"`
	ProjectID string `json:"project_id"`
}

type Project struct {
	ID    string `json:"id"`
	Name  string `json:"name"`
	Color string `json:"color"`
}

type Client struct {
	Token string
}

func NewClient(token string) *Client {
	return &Client{Token: token}
}

func (c *Client) GetTodayTasks() ([]Task, error) {
	req, _ := http.NewRequest("GET", apiBase+"/tasks?filter=today%20%7C%20overdue", nil)
	req.Header.Set("Authorization", "Bearer "+c.Token)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("todoist API error %d: %s", resp.StatusCode, string(body))
	}
	var tasks []Task
	err = json.NewDecoder(resp.Body).Decode(&tasks)
	return tasks, err
}

func (c *Client) GetTask(taskID string) (*Task, error) {
	req, _ := http.NewRequest("GET", apiBase+"/tasks/"+taskID, nil)
	req.Header.Set("Authorization", "Bearer "+c.Token)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("todoist API error %d", resp.StatusCode)
	}
	var task Task
	err = json.NewDecoder(resp.Body).Decode(&task)
	return &task, err
}

func (c *Client) GetProjects() ([]Project, error) {
	req, _ := http.NewRequest("GET", apiBase+"/projects", nil)
	req.Header.Set("Authorization", "Bearer "+c.Token)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("todoist API error %d", resp.StatusCode)
	}
	var projects []Project
	err = json.NewDecoder(resp.Body).Decode(&projects)
	return projects, err
}

func (c *Client) AddComment(taskID, content string) error {
	body := fmt.Sprintf(`{"task_id":"%s","content":"%s"}`, taskID, content)
	req, _ := http.NewRequest("POST", apiBase+"/comments", strings.NewReader(body))
	req.Header.Set("Authorization", "Bearer "+c.Token)
	req.Header.Set("Content-Type", "application/json")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	resp.Body.Close()
	return nil
}

func MatchProjectToCategory(projectID string, projects []Project, categories []db.Category) int {
	var projectName string
	for _, p := range projects {
		if p.ID == projectID {
			projectName = strings.ToLower(p.Name)
			break
		}
	}
	if projectName == "" {
		return -1
	}
	for i, cat := range categories {
		catName := strings.ToLower(cat.Title)
		if strings.Contains(projectName, catName) || strings.Contains(catName, projectName) {
			return i
		}
	}
	return -1
}

File: src/todoist.rs
use serde::{Deserialize, Serialize};

const TODOIST_API_BASE: &str = "https://api.todoist.com/rest/v2";

#[derive(Debug, Clone, Deserialize)]
pub struct TodoistTask {
    pub id: String,
    pub content: String,
    pub description: String,
    pub project_id: String,
    pub is_completed: bool,
    pub due: Option<TodoistDue>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct TodoistDue {
    pub date: String,
    pub string: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct TodoistProject {
    pub id: String,
    pub name: String,
    pub color: String,
}

#[derive(Debug, Serialize)]
struct TodoistComment {
    task_id: String,
    content: String,
}

pub struct TodoistClient {
    token: String,
}

impl TodoistClient {
    pub fn new(token: &str) -> Self {
        Self {
            token: token.to_string(),
        }
    }

    pub fn get_today_tasks(&self) -> anyhow::Result<Vec<TodoistTask>> {
        let resp: Vec<TodoistTask> = ureq::get(&format!("{}/tasks", TODOIST_API_BASE))
            .set("Authorization", &format!("Bearer {}", self.token))
            .query("filter", "today | overdue")
            .call()?
            .into_json()?;
        Ok(resp)
    }

    pub fn get_task(&self, task_id: &str) -> anyhow::Result<TodoistTask> {
        let resp: TodoistTask = ureq::get(&format!("{}/tasks/{}", TODOIST_API_BASE, task_id))
            .set("Authorization", &format!("Bearer {}", self.token))
            .call()?
            .into_json()?;
        Ok(resp)
    }

    pub fn get_projects(&self) -> anyhow::Result<Vec<TodoistProject>> {
        let resp: Vec<TodoistProject> = ureq::get(&format!("{}/projects", TODOIST_API_BASE))
            .set("Authorization", &format!("Bearer {}", self.token))
            .call()?
            .into_json()?;
        Ok(resp)
    }

    pub fn add_comment(&self, task_id: &str, content: &str) -> anyhow::Result<()> {
        let body = TodoistComment {
            task_id: task_id.to_string(),
            content: content.to_string(),
        };
        ureq::post(&format!("{}/comments", TODOIST_API_BASE))
            .set("Authorization", &format!("Bearer {}", self.token))
            .send_json(&body)?;
        Ok(())
    }

    /// Match a Todoist project name to a sesh category (case-insensitive substring match)
    pub fn match_project_to_category(
        &self,
        project_id: &str,
        projects: &[TodoistProject],
        categories: &[crate::db::categories::Category],
    ) -> Option<usize> {
        let project = projects.iter().find(|p| p.id == project_id)?;
        let project_name = project.name.to_lowercase();
        categories.iter().position(|c| {
            let cat_name = c.title.to_lowercase();
            project_name.contains(&cat_name) || cat_name.contains(&project_name)
        })
    }
}
